### ISSUE TITLE
Converting the output mesh to another format 

### ISSUE THREAD

**Issue by AlirezaSharif** [SVMTK/SVMTK#26]
Hi, 

I am trying to convert the output of SVMTK to two separate .xdmf files one of which contains the mesh with facet tags, and the other contains the mesh with cell tags. I have used "smap.add_interface" to have the facet tags inside the output file with .mesh. However, I could not find any straightforward way to do it based on my quick search. Is there any way to do that?

The way I have tried so far was to convert the output file to .msh format via Meshio, but I am getting this error: 
 KeyError: "Illegal ANSYS cell type 'line'. (legal: triangle, tetra, quad, hexahedron, pyramid, wedge)".
.msh allows me to extract the facet and cell tags easily via gmshio. 

Thank you in advance.

**Comment by larsmva** [comment]
Hi! 

You can do it with Meshio, try the following code.
```python
# meshio version 4.3.2
import meshio
# Read the .mesh file into meshio. 
mesh = meshio.read(meshfile)
# Extract subdomains and boundaries between regions into appropriate containers.
points = mesh.points
tetra  = {"tetra": mesh.cells_dict["tetra"]}
triangles = {"triangle": mesh.cells_dict["triangle"]} 
subdomains = {"subdomains": [mesh.cell_data_dict["medit:ref"]["tetra"]]}
boundaries = {"boundaries": [mesh.cell_data_dict["medit:ref"]["triangle"]]}
    
# Write just the mesh. 
xdmf = meshio.Mesh(points, tetra) 
meshio.write("mesh_only.xdmf", xdmf)
    
# Write the subdomains of the mesh.
xdmf = meshio.Mesh(points, tetra, cell_data=subdomains)
meshio.write("cell_tags.xdmf", xdmf)
    
# Write the boundaries/interfaces of the mesh.
xdmf = meshio.Mesh(points, triangles,cell_data=boundaries)
meshio.write("facet_tags.xdmf", xdmf)
```

**Comment by AlirezaSharif** [comment]
Thank you so much Lars!

Is it possible to save all the facets belonging to each subdomain to separate files (while keeping the tags for the interfaces)?
For cell tags, it is straightforward but since SVMTK does not label the internal facets, it is a bit tricky. Let me know your thoughts.

**Comment by larsmva** [comment]
It is possible, but you would need a mapping between the tetrahedrons and triangles. However, the .mesh format does not have that and I don't think there are any functions that does this in Meshio, so you would need to create the mapping yourself. 

What program do you intend to use the mesh with?

**Comment by AlirezaSharif** [comment]
I want to use the mesh in FEniCSx as my FE domain.

**Comment by larsmva** [comment]
I think you could that in FEniCSx, see [post](https://fenicsproject.discourse.group/t/transfer-meshtags-to-submesh-in-dolfinx/8952/7) .

**Comment by AlirezaSharif** [comment]
Thank you Lars!

I was trying the FEniCSx functionalities in the post you shared. Since we only have the interface facet tags, I have to save each subdomain cells with mesh tags of interface in .xdmf format. Then, I have to re-read this file and write it as a xdmf file containing facets. I am currently getting some errors in re-reading the mesh tags for the facets. If you have a better suggestion or software (like Salome) let me know. 
SVMTK is compatible with multiphenics which is designed for multi-domain systems. Thank you agian for your help.

**Comment by jorgensd** [comment]
@AlirezaSharif  if you post the code and errors you are Getting I might be able to help you

**Comment by AlirezaSharif** [comment]
Hi Jorgen, 

Here is the code: first I read the whole volumetric mesh and the facet tags and save the subdomain that I want.
``` python3
# cell_tags and facet_tags contain the information of the whole domain saved as cells and facet, respectively.
with XDMFFile(MPI.COMM_WORLD, "cell_tags.xdmf", "r") as xdmf:
    domain = xdmf.read_mesh(name="mesh")
    ct = xdmf.read_meshtags(domain, name="mesh")
domain.topology.create_connectivity(domain.topology.dim, domain.topology.dim - 1)
with XDMFFile(MPI.COMM_WORLD, "facet_tags.xdmf", "r") as xdmf:
    ft = xdmf.read_meshtags(domain, name="mesh")

tdim = domain.topology.dim
fdim = tdim - 1

# this function takes a mesh with its facet tags (ft), cell tags (ct), the lable of the cells that we want to keep as subdomain (lable), and
# name of the output file (output). The output of the function is the volumetric mesh of the subdomain in .xdmf format.
def sub_volume(domain, ft, ct, output, lable):
    c_to_f = domain.topology.connectivity(tdim, fdim)
    f_map = domain.topology.index_map(fdim)
    all_facets = f_map.size_local + f_map.num_ghosts
    # Create array with zeros for all facets that are not marked
    all_values = np.zeros(all_facets, dtype=np.int32)
    all_values[ft.indices] = ft.values

    subdomain, entity_map, _, _ = dolfinx.mesh.create_submesh(
        domain, tdim, ct.find(lable))

    subdomain.topology.create_entities(fdim)
    subf_map = subdomain.topology.index_map(fdim)
    subdomain.topology.create_connectivity(tdim, fdim)
    c_to_f_sub = subdomain.topology.connectivity(tdim, fdim)
    num_sub_facets = subf_map.size_local + subf_map.size_global
    sub_values = np.empty(num_sub_facets, dtype=np.int32)
    for i, entity in enumerate(entity_map):
        parent_facets = c_to_f.links(entity)
        child_facets = c_to_f_sub.links(i)
        for child, parent in zip(child_facets, parent_facets):
            sub_values[child] = all_values[parent]
    sub_meshtag = dolfinx.mesh.meshtags(subdomain, subdomain.topology.dim-1, np.arange(
        num_sub_facets, dtype=np.int32), sub_values)

    with dolfinx.io.XDMFFile(domain.comm, f"{output}.xdmf", "w") as xdmf:
        xdmf.write_mesh(subdomain)
        subdomain.topology.create_connectivity(fdim, tdim)
        xdmf.write_meshtags(sub_meshtag)

sub_volume(domain, ft, ct, "output", 1)
```
However, when I want to load the same subdomain with its facet tags via: 
```python3
with XDMFFile(MPI.COMM_WORLD, "output.xdmf", "r") as xdmf:
    sub_domain = xdmf.read_mesh(name="mesh")
sub_domain.topology.create_connectivity(sub_domain.topology.dim, sub_domain.topology.dim - 1)
with XDMFFile(MPI.COMM_WORLD, "output.xdmf", "r") as xdmf:
    facet_tag = xdmf.read_meshtags(sub_domain, name="mesh")
```

I get this error: RuntimeError: 
line 101, in <module>
    facet_tag = xdmf.read_meshtags(sub_domain, name="mesh")
  File "/miniconda3/envs/fenicsx-env/lib/python3.10/site-packages/dolfinx/io/utils.py", line 177, in read_meshtags
    return super().read_meshtags(mesh, name, xpath)
RuntimeError: Storage format "" is unknown

I am using FEniCSx 0.6.0.

**Comment by jorgensd** [comment]
I would need the actual mesh files to be able to debug this.

**Comment by AlirezaSharif** [comment]
I have just mailed you the mesh files.

**Comment by jorgensd** [comment]
Change the latter line to
```python
with XDMFFile(MPI.COMM_WORLD, "output.xdmf", "r") as xdmf:
    facet_tag = xdmf.read_meshtags(sub_domain, name="mesh_tags")
```
as, `print(sub_meshtag.name)` gives `mesh_tags`. You can rename it by calling `sub_meshtag.name = "add_text_here"`.

**Comment by AlirezaSharif** [comment]
Thank you so much Jorgen. It solved the error.
Be that as it may, after running the code for two different labels (1 and 2), I checked whether the subdomain meshes are still conforming or not and  I got some strange results. Apparently, the resulting subdomain has a different topology. 

![non_conf](https://github.com/SVMTK/SVMTK/assets/107373265/0b94fa18-b470-4de3-aa4e-18d61bf12613)
![conf](https://github.com/SVMTK/SVMTK/assets/107373265/846f2a4e-d850-45ef-8c22-00e37ab61e7e)

the upper image is the extracted subdomains that I have loaded in Paraview. the other one is the original conforming mesh. Am I missing something here?
